### DISCUSSION TITLE
GEN24_Ladesteuerung v0.11.0

### DISCUSSION THREAD

**Discussion by wiggal** [wiggal/GEN24_Ladesteuerung#35] [Announcements]
## ACHTUNG diese Version beinhaltet umfangreiche Änderungen in den config-Dateien!!!  
**Bitte CHANGELOG.md genau lesen und alle Anpassungen in den config-Dateien durchführen**  
**Oder: eigene Anpassungen in die mitgelieferten config-Dateien übernehen!**   

Änderung in html/index.php:  
- Navigation ohne Javascript und mehr responsive  

#### ACHTUNG bitte html/config.php anpassen!!!  
Änderung in html/config.php  
Die NavigationsTABs werden nun in dem Array `$TAB_config` konfiguriert (Tabname, Dateiname, Startauswahl, ein/aus-blenden)  
Die config.ini aus dem Pythonverzeichnis wir gelesen um auf die Variablen zugreifen zu können.  
Variable $SQLite_file  entfernt, sie wird nun aus der config.ini ermittelt  

**NEU:** 7_tab_Diagram.php (noch Prototyp)  
Damit werden die Daten aus der SQLite Datei graphisch aufbereitet.  
Erforderlich php-sqlite3:  
`sudo apt install php-sqlite3`  

Änderung in html/1_tab_LadeSteuerung.php  
- Am Ende der Seite wird der Anforderungszeitpunkt der Prognose ausgeben.  

Änderung in Solarprognose_WeatherData.py, Solcast_WeatherData.py, WeatherDataProvider2.py  
- Meldung eingefügt, wenn die definierten "dataAgeMaxInMinutes" noch nicht abgelaufen sind,  
  und deshalb keine Prognose angefordert wird.  

Änderung in SymoGen24Controller2.py  
- Logging wird nur geschrieben, wenn der Parameter "schreiben" übergeben wurde.  

### ACHTUNG umfangreiche Änderungen in config.ini, bitte anpassen!!!  
**NEU Zusatzkonfigurationen für die Ladeberechnung möglich (z.B. für Wintermonate usw.)  
- neu Variable `Zusatz_Ladebloecke` im Block `[Ladeberechnung]`  
- neue Blöcke `[Winter]`  und `[Uebergang]`  
- In den Blöcken `[Winter]`  und `[Uebergang]` zusätzliche Werte für bestimmte Monate definiert werden,  
  diese überschreiben die Werte im Block `[Ladeberechnung]`, im den entsprechenden Monaten.  
  Es können auch noch mehrere Blöcke definiert werden.  

** Kommentare in config.ini von "#" auf ";" umgestellt!!  
- die Kommentierung musste von "#" auf ";" umgestellt werden, um die Datei in PHP parsen zu können.  
  Dadurch kann man einfach in PHP auf die config.ini von Python zugreifen.  

<hr /><em>This discussion was created from the release <a href='https://github.com/wiggal/GEN24_Ladesteuerung/releases/tag/v0.11.0'>GEN24_Ladesteuerung v0.11.0</a>.</em>

**Comment by modezz** [comment]
Hallo,
funktioniert absolut top  - nur der scale_factor ist beim diagramm - sollte wohl eher /10 sein ist im code *10 - zur berechnung egal aber im diagramm eher suboptimal.

Beispiel:

Hatte 737,1 W PV Produktion --> 7371 im Diagramm

aktuellePrognose: 2551
RestTagesPrognose: 6141
PrognoseAbzugswert/Stunde: 200
TagesPrognose - Abzugswerte: 2176
Grundlast_Summe für Tag: 583
aktuellePVProduktion/Watt: 7371
aktuelleEinspeisung/Watt: -5
aktuelleBatteriePower/Watt: -640
GesamtverbrauchHaus/Watt: 6736
aktuelleBattKapazität/Watt: 2820
Batteriestatus in Prozent: 74.5 %
LadewertGrund: Prognoseberechnung / BatSparFaktor
Bisheriger Ladewert/Watt: 995
Bisheriger Ladewert/Prozent: 9.72 %
Neuer Ladewert/Watt: 1003
Neuer Ladewert/Prozent: 9.79 %
newPercent_schreiben: 0 

def get_mppt_power(self):
        mpppt_1_power = self.read_data('MPPT_1_DC_Power')
        mpppt_2_power = self.read_data('MPPT_2_DC_Power')
        power_scale_tmp = np.float64(np.int16(self.read_data('MPPT_Power_Scale_Factor')))
        return int((mpppt_1_power + mpppt_2_power) * (**10** ** power_scale_tmp))

**Reply by wiggal** [reply]
Hallo, bist du sicher?

Folgende Daten stehen doch im Protokoll:

```
aktuellePVProduktion/Watt: 7371 
aktuelleBatteriePower/Watt: -640
GesamtverbrauchHaus/Watt: 6736
```
`return int((mpppt_1_power + mpppt_2_power) * (10 ** power_scale_tmp`))

Bedeutet nicht Wert mpppt_X_power mal 10, sondern mpppt_X_power mal  (10 hoch power_scale_tmp),
wobei power_scale_tmp der Skalierungsfaktor vom Wechselrichter ist.

Mit den Werten, Produktion und Verbrauch hatte ich noch nie Fehler (kann aber natürlich trotzdem sein),
an der Berechnung habe ich schon ewig nicht mehr gedreht, und die Ergebnisse nur ins Diagramm übernommen.

Für weitere Hinweise zu einem Fehler bin ich aber jederzeit dankbar.
Schönen Gruß

**Comment by modezz** [comment]
lt fronius:

> Der Datentyp „sunssf“ ist ein signed integer mit 16bit.
> Rechenbeispiel:
> (Model 160): 1_DCW = 10000, DCW_SF = -1 -> Leistung = 10000 x 10^(-1) = 1000 W

würde passen - 7371 *10^(-1)=737,1W

**Reply by wiggal** [reply]
ist richtig, aber power_scale_tmp kann auch 0 oder -2 werden, und bei 0 wäre das Ergebnis dann 7371.

Aber warum denkst du, dass du nur 737Watt produziert hast?
Wenn du schon 640 Watt in die Batterie lädst und 5Watt einspeist, dann wäre dein Verbrauch nur 92Watt. Ist das plausibel?

Zu welchem Zeitpunkt war das, was sagt Solarweb zu dem Zeitpunkt.

Das wir uns nicht falsch verstehen, dass mal falsche Werte vom WR kommen ist nicht unmöglich, aber die Berechnung, denke ich passt.

Schönen Gruß

**Comment by modezz** [comment]
Servus Wiggal - die Berechung passt 100% - wie gesagt funktioniert TOP - nur im Diagramm ist es bescheiden anzusehen
![solarweb](https://github.com/wiggal/GEN24_Ladesteuerung/assets/38989365/2096732e-fd6d-4cd1-b762-ccc346c5e16e)
vs
![PVP](https://github.com/wiggal/GEN24_Ladesteuerung/assets/38989365/b1ca988c-b59a-4ec1-8124-fd2f9aefab06)

**Reply by wiggal** [reply]
Das sieht ja wirklich komisch aus. 
wie schon geschrieben, sind das die Werte die dein WR über Modbus liefert.
Sieht danach aus als würde der Faktor nicht passen. 
Ich würde mal Modbus aus und wieder einschalten. 
Und poste morgen bitte nochmal Daten.
Danke

**Reply by wiggal** [reply]
so sieht's heute bei mir aus:

![grafik](https://github.com/wiggal/GEN24_Ladesteuerung/assets/69506553/3218c7ce-5167-41be-a766-a55a0b34b665)

![grafik](https://github.com/wiggal/GEN24_Ladesteuerung/assets/69506553/19c30a16-8dcd-45f7-837f-81817d6c65cd)

**Comment by wiggal** [comment]
Hi,
hab mir nochmal die Doku der Register angesehen, und keinen Fehler gefunden.

Wenn du wieder mal den Verdacht hast, dass die Werte nicht stimmen, kannst du bitte mal die Ausgabe dieses Befehls posten:

`python3 SymoGen24Connector.py |grep MPPT`

und den zugehörigen Teil aus der Crontab.log

**Reply by modezz** [reply]
hi, danke für die Inputs - fyi habe nur einen MPPT aktiv - der 2te ist inaktiv speich da hängt nix dran - 
modbus wurde gestern neu gestartet
firmware ist die letzte verfügbare vom Gen24

40254: MPPT_SunspecID - 160.0
40256: MPPT_Current_Scale_Factor - 65532.0
40257: MPPT_Voltage_Scale_Factor - 65534.0
40258: MPPT_Power_Scale_Factor - 65534.0
40273: MPPT_1_DC_Current - 928.0
40274: MPPT_1_DC_Voltage - 50824.0
40275: MPPT_1_DC_Power - 4718.0
40280: MPPT_1_Temperature - 32768.0
40281: MPPT_1_State - 65535.0
40293: MPPT_2_DC_Current - 65535.0
40294: MPPT_2_DC_Voltage - 65535.0
40295: MPPT_2_DC_Power - 65535.0
40300: MPPT_2_Temperature - 32768.0
40301: MPPT_2_State - 65535.0

aktuellePrognose: 0
RestTagesPrognose: 0
PrognoseAbzugswert/Stunde: 300
TagesPrognose - Abzugswerte: 0
Grundlast_Summe für Tag: 0
aktuellePVProduktion/Watt: 709
aktuelleEinspeisung/Watt: 6
aktuelleBatteriePower/Watt: 171
GesamtverbrauchHaus/Watt: 874
aktuelleBattKapazität/Watt: 165
Batteriestatus in Prozent: 98.5 %
LadewertGrund: Stunde aus BattVollUm erreicht!!
Bisheriger Ladewert/Watt: 4999
Bisheriger Ladewert/Prozent: 48.82 %
Neuer Ladewert/Watt: 5000
Neuer Ladewert/Prozent: 48.82 %
newPercent_schreiben: 0

**Comment by modezz** [comment]
40254: MPPT_SunspecID - 160.0
40256: MPPT_Current_Scale_Factor - 65532.0
40257: MPPT_Voltage_Scale_Factor - 65534.0
40258: MPPT_Power_Scale_Factor - 65534.0
40273: MPPT_1_DC_Current - 686.0
40274: MPPT_1_DC_Voltage - 48753.0
40275: MPPT_1_DC_Power - 3345.0

![akt](https://github.com/wiggal/GEN24_Ladesteuerung/assets/38989365/b8c2b3cf-cf6b-4556-aa7b-e3eaf201a16c)

**Comment by modezz** [comment]
geht wohl in diese Richtung:

> 0 (SF = x10), 65535 (SF = 1), 65534 (SF = /10), 65533 (SF = /100)

**Reply by wiggal** [reply]
OK, es liegt wohl an dem nicht belegten MPPT2.
Vermutung zu  dem Beispiel vom 7.11.:
- in der Zeit an dem die Grafik unten ist, war der Skalierfaktor 10^-2, dann müssten die Werte ca. 655W zu hoch sein.
  Ist auch in dem Fall von heute mit `aktuellePVProduktion/Watt: 709` plausibel. 
  (65534+4718)/100 =702 Watt
  
- in der Zeit an dem die Grafik oben ist, war der Skalierfaktor 10^-1, dann müssten die Werte ca. 6553W zu hoch sein.

Das wäre noch zu verifizieren, kannst du mir bitte noch bei folgenden Bedingungen die Daten zu den Befehlen liefern.
Wenn die Produktion weit über 1000 Watt ist (dann müsste der Skalierfaktor 10^-1 und der Produktionswert 6553Watt zu hoch sein.)

```
python3 SymoGen24Connector.py |grep MPPT

python3 SymoGen24Controller2.py

```

Wenn dem so ist, dann muss ich für den Fall, dass der MPPT2 deaktiviert ist, also wenn er `40295: MPPT_2_DC_Power - 65535.0` den Wert auf Null setzen.

SymoGen24Connector.py müsste dann ab Zeile 270 so aussehen:

```
    def get_mppt_power(self):
        mpppt_1_power = self.read_data('MPPT_1_DC_Power')
        mpppt_2_power = self.read_data('MPPT_2_DC_Power')
        if mpppt_2_power == 65535:
            mpppt_2_power = 0
        power_scale_tmp = np.float64(np.int16(self.read_data('MPPT_Power_Scale_Factor')))
        return int((mpppt_1_power + mpppt_2_power) * (10 ** power_scale_tmp))

```
Schönen Gruß Wiggal

**Reply by modezz** [reply]
ok -  ich werde mal den code ändern im connector.py dann wissen wir morgen schon mehr, habe leider untertags wenig möglichkeit das zu testen wenn die produktion >1000W ist

Danke!
SG

**Reply by wiggal** [reply]
hast du dann heute den Sprung auch als die Produktion an die 1000Watt ging?

wenn du den Code angepasst hast, zum Test den Befehl manuell ausführen.

**Comment by modezz** [comment]
lt Solarweb 
09:25 - 783W zur Batterie und 205W direkt verbraucht
09:30 - 1305W zur Batterie und 205W direkt verbraucht

************* BEGINN:  2023-11-08 09:25:02.304248 ************* 

######### L A D E S T E U E R U N G #########

aktuellePrognose:            5069
RestTagesPrognose:           25378
PrognoseAbzugswert/Stunde:   300
TagesPrognose - Abzugswerte: 4826
Grundlast_Summe fÃ¼r Tag:     1375
aktuellePVProduktion/Watt:   7688
aktuelleEinspeisung/Watt:    4
aktuelleBatteriePower/Watt:  -924
GesamtverbrauchHaus/Watt:    6760
aktuelleBattKapazitÃ¤t/Watt:  7111
Batteriestatus in Prozent:   35.7 %
LadewertGrund:  BattStatusProz < MindBattLad
Bisheriger Ladewert/Watt:    4999
Bisheriger Ladewert/Prozent: 48.82 %
Neuer Ladewert/Watt:         5000
Neuer Ladewert/Prozent:      48.82 %
newPercent_schreiben:        0

Alte und Neue Werte unterscheiden sich weniger als die Schreibgrenzen des WR, NICHTS zu schreiben!!


************* ENDE:  2023-11-08 09:25:02.323491 ************* 

************* BEGINN:  2023-11-08 09:30:02.789768 ************* 

######### L A D E S T E U E R U N G #########

aktuellePrognose:            5160
RestTagesPrognose:           24993
PrognoseAbzugswert/Stunde:   300
TagesPrognose - Abzugswerte: 4754
Grundlast_Summe fÃ¼r Tag:     1350
aktuellePVProduktion/Watt:   8601
aktuelleEinspeisung/Watt:    0
aktuelleBatteriePower/Watt:  -1826
GesamtverbrauchHaus/Watt:    6775
aktuelleBattKapazitÃ¤t/Watt:  7000
Batteriestatus in Prozent:   36.7 %
LadewertGrund:  BattStatusProz < MindBattLad
Bisheriger Ladewert/Watt:    4999
Bisheriger Ladewert/Prozent: 48.82 %
Neuer Ladewert/Watt:         5000
Neuer Ladewert/Prozent:      48.82 %
newPercent_schreiben:        0

**Comment by modezz** [comment]
wiggal,das schaut guad aus! SG

ohne codeänderung:

######### L A D E S T E U E R U N G #########

aktuellePrognose:            0
RestTagesPrognose:           0
PrognoseAbzugswert/Stunde:   300
TagesPrognose - Abzugswerte: 0
Grundlast_Summe für Tag:     0
aktuellePVProduktion/Watt:   **656**

mit adaptierten code:

######### L A D E S T E U E R U N G #########

aktuellePrognose:            0
RestTagesPrognose:           0
PrognoseAbzugswert/Stunde:   300
TagesPrognose - Abzugswerte: 0
Grundlast_Summe für Tag:     0
aktuellePVProduktion/Watt:   **0**

**Reply by wiggal** [reply]
sieht gut aus, mich wundert nur, dass der MPPT2 nicht 0 liefert.
ich werd es trotzdem für beide MPPT einbauen, 65535 kommt ja normal nicht vor.
Vielleicht schreibst du morgen abend nochmal.
Danke dir
Wiggal

**Comment by wiggal** [comment]
Hi, habs mal in den Issues dokumentiert. https://github.com/wiggal/GEN24_Ladesteuerung/issues/38

**Comment by modezz** [comment]
es funktioniert nun 1A-Spitzenklasse

************* BEGINN: 2023-11-09 10:50:02.722789 *************

######### L A D E S T E U E R U N G #########

aktuellePrognose: 1576
RestTagesPrognose: 3834
PrognoseAbzugswert/Stunde: 300
TagesPrognose - Abzugswerte: 0
Grundlast_Summe für Tag: 950
aktuellePVProduktion/Watt: 4238
aktuelleEinspeisung/Watt: 2672
aktuelleBatteriePower/Watt: -1332 

aktuell am Abend:

40254: MPPT_SunspecID - 160.0
40256: MPPT_Current_Scale_Factor - 65532.0
40257: MPPT_Voltage_Scale_Factor - 65534.0
40258: MPPT_Power_Scale_Factor - 65534.0
40273: MPPT_1_DC_Current - 263.0
40274: MPPT_1_DC_Voltage - 26152.0
40275: MPPT_1_DC_Power - 687.0
40280: MPPT_1_Temperature - 32768.0
40281: MPPT_1_State - 65535.0
40293: MPPT_2_DC_Current - 65535.0
40294: MPPT_2_DC_Voltage - 65535.0
40295: MPPT_2_DC_Power - 65535.0
40300: MPPT_2_Temperature - 32768.0
40301: MPPT_2_State - 65535.0

![akt](https://github.com/wiggal/GEN24_Ladesteuerung/assets/38989365/67c4a72d-c6a9-476c-8cfc-71c334fff2f9)

**Reply by wiggal** [reply]
OK,
super, dann werd ich am Wochenende eine neue Version ziehen.

Übrigens das Diagramm ist nur ein Prototyp, hab einfach die Daten, die ich schon hatte in das SQlite-File geschrieben, hat aber für den Fall gute Dienste geleistet. ;-)
Ich werde die Datengrundlage nochmal von Grund auf neu gestalten, wenn ich Zeit finde.

Schönen Gruß
Wiggal

**Reply by modezz** [reply]
Super danke nochmals!!

**Comment by wiggal** [comment]
erl.
